source/generators/LambertianGenerator.cc

```diff
@@ -0,0 +1,150 @@
+#include <G4Event.hh>
+#include <G4RandomDirection.hh>
+#include <Randomize.hh>
+#include <G4RandomTools.hh>
+#include <G4OpticalPhoton.hh>
+
+#include "CLHEP/Units/SystemOfUnits.h"
+
+using namespace nexus;
+using namespace CLHEP;
+
+REGISTER_CLASS(LED450LWcone, G4VPrimaryGenerator)
+
+
+LED450LWcone::LED450LWcone():
+G4VPrimaryGenerator(), msg_(0), particle_definition_(0),
+energy_min_(0.), energy_max_(0.), geom_(0), momentum_{},
+costheta_min_(-1.), costheta_max_(1.), phi_min_(0.), phi_max_(2.*pi)
+{
+  msg_ = new G4GenericMessenger(this, "/Generator/LED450LWcone/",
+    "Control commands of single-particle generator.");
+
+  msg_->DeclareMethod("particle", &LED450LWcone::SetParticleDefinition,
+    "Set particle to be generated.");
+
+  G4GenericMessenger::Command& min_energy =
+    msg_->DeclareProperty("min_energy", energy_min_,
+      "Set minimum kinetic energy of the particle.");
+  min_energy.SetUnitCategory("Energy");
+  min_energy.SetParameterName("min_energy", false);
+  min_energy.SetRange("min_energy>0.");
+
+  G4GenericMessenger::Command& max_energy =
+    msg_->DeclareProperty("max_energy", energy_max_,
+      "Set maximum kinetic energy of the particle");
+  max_energy.SetUnitCategory("Energy");
+  max_energy.SetParameterName("max_energy", false);
+  max_energy.SetRange("max_energy>0.");
+
+  msg_->DeclareProperty("region", region_,
+    "Set the region of the geometry where the vertex will be generated.");
+
+
+  msg_->DeclarePropertyWithUnit("momentum", "mm",  momentum_, "Set particle 3-momentum.");
+
+  msg_->DeclareProperty("min_costheta", costheta_min_,
+			"Set minimum cosTheta for the direction of the particle.");
+  msg_->DeclareProperty("max_costheta", costheta_max_,
+			"Set maximum cosTheta for the direction of the particle.");
+  msg_->DeclareProperty("min_phi", phi_min_,
+			"Set minimum phi for the direction of the particle.");
+  msg_->DeclareProperty("max_phi", phi_max_,
+			"Set maximum phi for the direction of the particle.");
+
+
+  DetectorConstruction* detconst = (DetectorConstruction*) G4RunManager::GetRunManager()->GetUserDetectorConstruction();
+  geom_ = detconst->GetGeometry();
+}
+
+
+
+LED450LWcone::~LED450LWcone()
+{
+  delete msg_;
+}
+
+
+
+void LED450LWcone::SetParticleDefinition(G4String particle_name)
+{
+  particle_definition_ =
+    G4ParticleTable::GetParticleTable()->FindParticle(particle_name);
+
+  if (!particle_definition_)
+    G4Exception("[LED450LW]", "SetParticleDefinition()",
+      FatalException, "User gave an unknown particle name.");
+}
+
+
+
+void LED450LWcone::GeneratePrimaryVertex(G4Event* event)
+{
+  // Generate uniform random energy in [E_min, E_max]
+  G4double kinetic_energy = nexus::UniformRandomInRange(energy_max_, energy_min_);
+
+  // Calculate cartesian components of momentum
+  G4double mass   = particle_definition_->GetPDGMass();
+  G4double energy = kinetic_energy + mass;
+  G4double pmod = std::sqrt(energy*energy - mass*mass);
+
+  G4ThreeVector dir = momentum_;
```

Comment:
This assignment is not needed, you can just use the `momentum_` variable, right?